src/cocotb/regression.py

```diff
@@ -165,132 +218,81 @@ def __init__(self) -> None:
         package_name = os.getenv("RESULT_TESTPACKAGE", "all")
 
         self.xunit = XUnitReporter(filename=results_filename)
-
         self.xunit.add_testsuite(name=suite_name, package=package_name)
-
         self.xunit.add_property(name="random_seed", value=str(cocotb.RANDOM_SEED))
 
-    def discover_tests(
-        self, modules: Iterable[str], filters: Optional[Sequence[str]] = None
-    ) -> None:
+    def discover_tests(self, modules: Sequence[str]) -> None:
         """Discover tests in files automatically.
 
         Should be called before :meth:`start_regression` is called.
-        If no filters are given, all tests are included.
-        If one or more filters are given, only those tests which match a filter are included; the rest are excluded.
 
         Args:
-            modules: a list of module names to execute to find tests.
-            filters: regex patterns for test names.
-                A match *includes* the test.
+            modules: A list of module names to execute to find tests.
 
+        Raises:
+            RuntimeError: If no tests are found.
         """
         for module_name in modules:
             self.log.debug("Searching for tests in module %s", module_name)
             self.log.debug("Python Path: %s", ",".join(sys.path))
             self.log.debug("PWD: %s", os.getcwd())
-            __import__(module_name)
+            mod = import_module(module_name)
+
+            if not hasattr(mod, "__cocotb_tests__"):
+                raise RuntimeError(f"No tests were discovered in module: {module_name}")
 
-        self.ntests = len(self._queue)
+            for test in mod.__cocotb_tests__:
+                self.register_test(test)
 
         # error if no tests were discovered
-        if self.ntests == 0:
+        if not self._test_queue:
             modules_str = ", ".join(repr(m) for m in modules)
-            raise RuntimeError(f"No tests were discovered in modules: {modules_str}")
+            raise RuntimeError(f"No tests were discovered in any module: {modules_str}")
 
-        # if no filters, accept all tests
-        if not filters:
-            return
+    def filter_tests(self, filters: Sequence[str]) -> None:
+        """Filters discovered tests.
 
-        # else mark all tests as excluded
-        self.ntests = 0
-        for test in self._queue:
-            test.included = False
+        Only those tests which match at least one of the given filters are included;
```

Comment:
```suggestion

        Only those tests which match at least one of the given filters are included;
```

Reply:
I'm not sure adding an additional newline here is required by PEP257.

Reply:
There is already a blank line there.